Add native support for tile padding in generic reduce OPs

varshith257 opened the pull request:
### Ticket
Resolves #25549

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212832436210978

varshith257 committed:
initial changes

varshith257 committed:
add initial changes

minaliuTT commented:
@varshith257 Can you please look into this? fyi @marty1885

varshith257 committed:
Merge branch 'main' into feat/tile-padding-reduce

varshith257 committed:
Fix tile padding for reduce ops: use shared pad_tile.hpp

- Remove duplicate pad_tile.hpp from reduction/generic/device/kernels/dataflow
- Extend shared kernel_helper_functions/pad_tile.hpp with:
  - NeutralPolicy enum (Zero, NegInf, PosInf) for different reduce ops
  - get_neutral_value() template to get fill values for Float32/Bfloat16
  - pad_last_wtile() for width dimension padding in reduction
  - pad_last_htile() for height dimension padding in reduction
- Update reader kernels to include shared pad_tile.hpp
- Update reduce_op_multi_core_h_program_factory.cpp with:
  - Calculation of padding dimensions from logical shape
  - Pass IN_DF, LAST_W, LAST_H, NEUTRAL_KIND compile-time args
  - Pass is_last_w_padded, is_last_h_padded runtime args for sharded path

Addresses reviewer feedback to use shared pad_tile.hpp instead of duplicate.

varshith257 committed:
Add native tile padding for W and HW reduction ops

- Create reader_unary_reduce_w_with_padding.cpp for W-reduction
  with tile position tracking and padding support
- Update reduce_op_multi_core_w_program_factory.cpp:
  - Add get_neutral_policy() helper
  - Calculate padding dimensions from logical shape
  - Pass Wt, Ht, IN_DF, LAST_W, LAST_H, NEUTRAL_KIND compile-time args
  - Use new padding-aware reader kernel
- Update reduce_op_single_core_hw_program_factory.cpp:
  - Add get_neutral_policy() helper
  - Calculate padding dimensions from logical shape
  - Pass padding compile-time args to reader
  - Use padding-aware reader kernel for HW reduction

This completes native tile padding support for all generic reduction
operations (H, W, HW dimensions), eliminating the need for the
separate fill_implicit_tile_padding preprocessing step.

varshith257 committed:
Add padding-specific unit test for reduce operations

Add test_reduce_with_padding to verify that reduction operations
(sum, max, min, mean) work correctly when the logical shape differs
from the padded shape. Tests cover:
- W-dimension reduction (dim=-1)
- H-dimension reduction (dim=-2)
- HW-reduction (dim=None, full reduce)
- Various padding scenarios: W-only, H-only, both H and W

This test validates that native tile padding in reduction kernels
produces correct results matching PyTorch reference.

varshith257 committed:
Add padding-specific unit test for reduce operations

Add test_reduce_with_padding to verify that reduction operations
(sum, max, min, mean) work correctly when the logical shape differs
from the padded shape. Tests cover:
- W-dimension reduction (dim=-1)
- H-dimension reduction (dim=-2)
- HW-reduction (dim=None, full reduce)
- Various padding scenarios: W-only, H-only, both H and W

This test validates that native tile padding in reduction kernels
produces correct results matching PyTorch reference.

bbradelTT commented on ttnn/cpp/ttnn/operations/reduction/generic/device/reduce_op_single_core_hw_program_factory.cpp:
```diff
@@ -87,7 +120,7 @@ ReduceSingleCoreHwProgramFactory::cached_program_t ReduceSingleCoreHwProgramFact
         program,
         "ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/dataflow/"
-        "reader_unary_reduce_universal_start_id.cpp",
+        "reader_unary_reduce_w_with_padding.cpp",
```
since this is only for generic reduce, the new kernel should be in the reduction/generic subdirectory.

varshith257 replied:
done

varshith257 committed:
Merge branch 'feat/tile-padding-reduce' of https://github.com/varshith257/tt-metal into feat/tile-padding-reduce

varshith257 committed:
Merge branch 'main' into feat/tile-padding-reduce

varshith257 commented on ttnn/cpp/ttnn/operations/reduction/generic/device/reduce_op_single_core_hw_program_factory.cpp:
```diff
@@ -78,7 +99,19 @@ ReduceSingleCoreHwProgramFactory::cached_program_t ReduceSingleCoreHwProgramFact
+
+    // Prepare reader compile-time args with padding support
+    uint32_t in_df = static_cast<uint32_t>(src0_cb_data_format);
+    uint32_t neutral_kind = get_neutral_policy(operation_attributes.math_op);
```
okay

jberkowitzTT commented:
@varshith257 - please update the branch and resolve the CI errors. Let us know if you have any questions or need support

Thanks!

jberkowitzTT commented:
@varshith257 - can you confirm that you are still working on this issue? Thank you!

varshith257 committed:
fix: padding template parameters to remove namespace prefix for DataFormat

varshith257 commented:
> Thank you for all of your efforts on this bounty, and let me know if there's anything I can do to help!

@jberkowitzTT Thanks for reminding. Sorry, I got bit busy with some other works last two weeks. Now, I am back again and will push this to complete in next 2 days if everything goes well.

varshith257 commented:
@bbradelTT @jberkowitzTT @minaliuTT @Aswinmcw Can you please trigger CI?

varshith257 committed:
Merge branch 'main' of https://github.com/tenstorrent/tt-metal into feat/tile-padding-reduce

varshith257 committed:
fix:Float16_b in padding functions

varshith257 committed:
fix: add implicit tile padding for non-H/W dimension reductions

varshith257 committed:
fix: fill padding after transpose in reduce_nd_loop

The transpose operation creates a new output tensor whose padding regions
are uninitialized. This caused garbage output for non-H/W dimension
reductions (like max(dim=-3) on a 3D tensor) because the H-reduce kernel
would read uninitialized padding values instead of the neutral value (-inf
for max, +inf for min, 0 for sum).

This fix adds fill_implicit_tile_padding calls after each transpose in the
reduce_nd_loop to ensure the new tensor's padding regions are filled with
the correct neutral value for the reduction operation.

varshith257 commented:
@bbradelTT @jberkowitzTT @minaliuTT @Aswinmcw Can you please trigger CI?

varshith257 committed:
Merge branch 'main' of https://github.com/tenstorrent/tt-metal into feat/tile-padding-reduce

varshith257 committed:
fix: use cb_in0.get_write_ptr() for native padding in H-reduce kernels

The H-reduce reader kernels (interleaved and sharded) use the experimental
circular buffer API (cb_in0.reserve_back/push_back) but the native tile
padding code referenced an undefined variable 'l1_write_addr'. This caused
undefined behavior, producing garbage output values.

Fix: obtain the L1 write address via cb_in0.get_write_ptr() before calling
apply_width_padding/apply_height_padding.

varshith257 commented:
@bbradelTT @jberkowitzTT @Aswinmcw Can you please trigger CI?

varshith257 committed:
Merge branch 'main' of https://github.com/tenstorrent/tt-metal into feat/tile-padding-reduce

jberkowitzTT committed:
Merge branch 'main' into feat/tile-padding-reduce

varshith257 committed:
fix: remove unused includes in generic reductions and reduce op single core factory

varshith257 committed:
fix: fill implicit tile padding after tensor transposition in reduce_impl

varshith257 committed:
Merge branch 'feat/tile-padding-reduce' of https://github.com/varshith257/tt-metal into feat/tile-padding-reduce

varshith257 committed:
Merge branch 'main' of https://github.com/tenstorrent/tt-metal into feat/tile-padding-reduce

varshith257 committed:
fix: replace TILE_WIDTH and TILE_HEIGHT with tt::constants in reduction factories

varshith257 committed:
Merge branch 'feat/tile-padding-reduce' of https://github.com/varshith257/tt-metal; branch 'main' of https://github.com/tenstorrent/tt-metal into feat/tile-padding-reduce

jberkowitzTT committed:
Merge branch 'main' into feat/tile-padding-reduce

bbradelTT reviewed (commented):
If you get all of this sorted out and the tests that were listed passing then please run this test with tracy to see the perf differences are. You can post the generated csv file.

Easiest command is
```
python -m tracy -r -m pytest tests/ttnn/unit_tests/operations/reduce/test_reduction.py::test_reduce_with_padding
```

jberkowitzTT committed:
Merge branch 'main' into feat/tile-padding-reduce

jberkowitzTT committed:
Merge branch 'main' into feat/tile-padding-reduce

varshith257 committed:
fix: review comments

varshith257 committed:
Merge branch 'feat/tile-padding-reduce' of https://github.com/varshith257/tt-metal into feat/tile-padding-reduce

varshith257 committed:
Merge branch 'main' of https://github.com/tenstorrent/tt-metal into feat/tile-padding-reduce

jberkowitzTT committed:
Merge branch 'main' into feat/tile-padding-reduce

jberkowitzTT committed:
Merge branch 'main' into feat/tile-padding-reduce

jberkowitzTT committed:
Merge branch 'main' into feat/tile-padding-reduce

jberkowitzTT committed:
Merge branch 'main' into feat/tile-padding-reduce
